shiny/ui/_chat.py

```diff
@@ -0,0 +1,900 @@
+
+    def set_user_message(self, value: str):
+        """
+        Set the user's message.
+
+        Parameters
+        ----------
+        value
+            The value to set the user input to.
+        """
+
+        _utils.run_coro_sync(
+            self._session.send_custom_message(
+                "shinyChatMessage",
+                {
+                    "id": self.id,
+                    "handler": "shiny-chat-set-user-input",
+                    "obj": value,
+                },
+            )
+        )
+
+    async def clear_messages(self):
+        """
+        Clear all chat messages.
+        """
+        self._messages.set(())
+        await self._send_custom_message("shiny-chat-clear-messages", None)
+
+    def destroy(self):
+        """
+        Destroy the chat instance.
+        """
+        for x in self._effects:
+            x.destroy()
+        self._effects.clear()
+
+    async def _remove_loading_message(self):
+        await self._send_custom_message("shiny-chat-remove-loading-message", None)
+
+    async def _send_custom_message(self, handler: str, obj: ClientMessage | None):
+        await self._session.send_custom_message(
+            "shinyChatMessage",
+            {
+                "id": self.id,
+                "handler": handler,
+                "obj": obj,
+            },
+        )
+
+
+@add_example(ex_dir="../api-examples/chat")
+def chat_ui(
+    id: str,
+    *,
+    placeholder: str = "Enter a message...",
+    width: CssUnit = "min(680px, 100%)",
+    height: CssUnit = "auto",
+    fill: bool = True,
+    **kwargs: TagAttrValue,
+) -> Tag:
+    """
+    UI container for a chat component (Shiny Core).
+
+    This function is for locating a :class:`~shiny.ui.Chat` instance in a Shiny Core
+    app. If you are using Shiny Express, use the :method:`~shiny.ui.Chat.ui` method
+    instead.
+
+    Parameters
+    ----------
+    id
+        A unique identifier for the chat UI.
+    placeholder
+        Placeholder text for the chat input.
+    width
+        The width of the chat container.
+    height
+        The height of the chat container.
+    fill
+        Whether the chat should vertically take available space inside a fillable container.
+    kwargs
+        Additional attributes for the chat container element.
+    """
+
+    id = resolve_id(id)
+
+    res = Tag(
+        "shiny-chat-container",
+        chat_deps(),
+        {
+            "style": css(
+                width=as_css_unit(width),
+                height=as_css_unit(height),
+            )
+        },
+        id=id,
+        placeholder=placeholder,
+        fill=fill,
+        **kwargs,
+    )
+
+    if fill:
+        res = as_fillable_container(as_fill_item(res))
+
+    return res
+
+
+def _express_is_active() -> bool:
+    from ..express._run import get_top_level_recall_context_manager
+
+    try:
+        get_top_level_recall_context_manager()
+        return True
+    except RuntimeError:
+        return False
```

Comment:
@wch this apparently will return `True` in some Shiny Core situations. Any idea how we do this more accurately?

Reply:
This function is used from `chat.ui()`. The case where it could be problematic is when `chat.ui()`is used in a function marked with `@expressify` or `@shiny.express.module` AND that function/module is used in a Shiny Core app. In that case, this `_express_is_active()` function will return `False`, and `chat.ui()` will see that value and raise an error, but that would be incorrect behavior.

It _should_ be possible to use `chat.ui()` from a function/module that's decorated with `@expressify` or `@shiny.express.module`, even if that function/module is used in a Shiny Core app.

Reply:
After some discussion, @cpsievert and I decided just to not use this check.